File: req-rep/cliente.py
import zmq
import datetime
import threading
import json
import sys
import time
import getpass # Para esconder a senha

def message_listener(sub_socket, user_name):
    """Função que roda em background para ouvir mensagens do Pub/Sub."""
    print(f"\n✅ Audição iniciada para o usuário '{user_name}'. Você receberá mensagens aqui.")
    while True:
        try:
            full_message = sub_socket.recv_string()
            topic, payload = full_message.split(' ', 1)
            message_data = json.loads(payload)
            sys.stdout.write('\r' + ' ' * 80 + '\r')
            if 'from' in message_data:
                print(f"📩 [Mensagem de {message_data['from']}]: {message_data['message']}")
            else:
                print(f"📢 [{topic}] {message_data['user']}: {message_data['message']}")
            sys.stdout.write(f"[{user_name}] Digite uma opção: ")
            sys.stdout.flush()
        except zmq.ZMQError as e:
            print(f"\n[ERRO DE AUDIÇÃO] Conexão perdida: {e}")
            break

def main_app(user_nome, session_token, req_socket, sub_socket):
    """Função principal da aplicação, executada após o login."""
    sub_socket.setsockopt_string(zmq.SUBSCRIBE, user_nome)
    listener_thread = threading.Thread(target=message_listener, args=(sub_socket, user_nome), daemon=True)
    listener_thread.start()
    time.sleep(0.5)
    
    opcao = ""
    while opcao != "sair":
        opcao = input(f"[{user_nome}] Digite uma opção (ou 'ajuda'): ").lower()

        match opcao:
            case "ajuda":
                print("\n--- MENU DE OPÇÕES ---\n"
                      "listar_user      - Lista todos os usuários\n"
                      "listar_canais    - Lista todos os canais\n"
                      "add_canal        - Adiciona um novo canal\n"
                      "inscrever [canal] - Inscreve você em um canal\n"
                      "publicar [canal]  - Publica uma mensagem em um canal\n"
                      "mensagem [user]   - Envia uma mensagem direta para um usuário\n"
                      "sair             - Desloga e encerra o cliente\n"
                      "----------------------")
            
            case "add_canal":
                titulo = input("Entre com o nome do canal: ")
                desc = input("Entre com a descrição do canal: ")
                request = {"service": "addChannel", "data": {"token": session_token, "titulo": titulo, "desc": desc}}
                req_socket.send_json(request)
                reply = req_socket.recv_json()
                if reply.get("status") == "OK":
                    print("\n✅ Canal adicionado com sucesso!")
                else:
                    print(f"\n❌ {reply.get('message')}")

            case cmd if cmd.startswith("inscrever"):
                try:
                    _, nome_canal = cmd.split()
                    sub_socket.setsockopt_string(zmq.SUBSCRIBE, nome_canal.lower())
                    print(f"✅ Inscrito com sucesso no canal '{nome_canal.lower()}'")
                except ValueError:
                    print("❌ Comando inválido. Use: inscrever <nome_do_canal>")

            case cmd if cmd.startswith("publicar"):
                try:
                    _, nome_canal = cmd.split()
                    msg_para_enviar = input(f"Mensagem para o canal '{nome_canal}': ")
                    timestamp = datetime.datetime.now().isoformat()
                    request = {"service": "publish", "data": {"token": session_token, "channel": nome_canal, "message": msg_para_enviar, "timestamp": timestamp}}
                    req_socket.send_json(request)
                    reply = req_socket.recv_json()
                    if reply.get('data', {}).get('status') == 'OK':
                        print("✅ Mensagem enviada para o servidor.")
                    else:
                        print(f"❌ Erro: {reply.get('data', {}).get('message')}")
                except ValueError:
                    print("❌ Comando inválido. Use: publicar <nome_do_canal>")

            case cmd if cmd.startswith("mensagem"):
                try:
                    _, user_destino = cmd.split()
                    msg_para_enviar = input(f"Mensagem para '{user_destino}': ")
                    timestamp = datetime.datetime.now().isoformat()
                    request = {"service": "message", "data": {"token": session_token, "dst": user_destino, "message": msg_para_enviar, "timestamp": timestamp}}
                    req_socket.send_json(request)
                    reply = req_socket.recv_json()
                    if reply.get('status') == 'OK':
                        print("✅ Mensagem enviada para o servidor.")
                    else:
                        print(f"❌ Erro: {reply.get('message')}")
                except ValueError:
                    print("❌ Comando inválido. Use: mensagem <usuario_destino>")

            case "listar_user":
                req_socket.send_json({"service": "listUsers", "data": {}})
                users = req_socket.recv_json()
                print("\n--- Usuários Cadastrados ---")
                for uid, uinfo in users.items():
                    print(f" -> {uinfo['user']}")
                print("----------------------------")
            
            case "listar_canais":
                req_socket.send_json({"service": "listChannels", "data": {}})
                canais = req_socket.recv_json()
                print("\n--- Canais Disponíveis ---")
                for cid, cinfo in canais.items():
                    print(f" -> {cinfo['titulo']}")
                print("------------------------")
                
            case "sair":
                print("Deslogando e encerrando...")
                break
                
            case _:
                if opcao:
                    print("Opção inválida. Digite 'ajuda' para ver os comandos.")

if __name__ == "__main__":
    context = zmq.Context()
    req_socket = context.socket(zmq.REQ)
    req_socket.connect("tcp://localhost:5555")
    
    while True:
        print("\n--- BEM-VINDO ---")
        print("1. Login")
        print("2. Cadastrar novo usuário")
        print("3. Sair")
        choice = input("Escolha uma opção: ")

        if choice == '1':
            user = input("Usuário: ")
            senha = getpass.getpass("Senha: ") # getpass esconde a senha
            request = {"service": "login", "data": {"user": user, "senha": senha}}
            req_socket.send_json(request)
            reply = req_socket.recv_json()
            if reply.get("status") == "OK":
                print("✅ Login realizado com sucesso!")
                # Inicia a aplicação principal após o login
                sub_socket = context.socket(zmq.SUB)
                sub_socket.connect("tcp://localhost:5557")
                main_app(reply.get("user"), reply.get("token"), req_socket, sub_socket)
                break # Sai do loop de login/cadastro
            else:
                print(f"❌ Falha no login: {reply.get('message')}")

        elif choice == '2':
            user = input("Digite o novo nome de usuário: ")
            senha = getpass.getpass("Digite a nova senha: ")
            request = {"service": "addUser", "data": {"user": user, "senha": senha}}
            req_socket.send_json(request)
            reply = req_socket.recv_string()
            if reply == "OK":
                print("\n✅ Usuário cadastrado com sucesso! Agora você pode fazer o login.")
            else:
                print(f"\n❌ {reply}")
        
        elif choice == '3':
            break
        
        else:
            print("Opção inválida.")
            
    print("\nPrograma encerrado. Até logo!")

File: req-rep/cliente.js
// cliente_completo.js (CommonJS) - versão compatível com servidor corrigido
const zmq = require("zeromq");
const prompt = require("prompt-sync")({ sigint: true });
const { setTimeout } = require("timers/promises");

// Endereços
const REQ_ADDR = "tcp://broker:5555";
const SUB_ADDR = "tcp://proxy:5557";
let subSocket = null;

// Lamport clock
let lamportClock = 0;
function incClock() { lamportClock++; return lamportClock; }
function updateClock(received) { lamportClock = Math.max(lamportClock, Number(received) || 0) + 1; return lamportClock; }

// Estado do cliente
let currentUser = null;
let sessionToken = null;
let isBot = false;

// Listener assíncrono de mensagens
async function messageListener(userName, autoSubscribe = false) {
  subSocket = new zmq.Subscriber();
  subSocket.connect(SUB_ADDR);
  subSocket.subscribe(userName); // sempre recebe DMs

  if (autoSubscribe) {
    console.log("🤖 Bot detectado: inscrito em todos os canais por padrão.");
  }

  for await (const [topicBuf, msgBuf] of subSocket) {
    try {
      const topic = topicBuf.toString();
      const payload = JSON.parse(msgBuf.toString());
      const receivedLamport = payload.lamport_clock || 0;
      updateClock(receivedLamport);

      process.stdout.write('\r' + ' '.repeat(80) + '\r');
      if (payload.from) {
        console.log(`📩 [DM de ${payload.from}]: ${payload.message}`);
      } else if (payload.user) {
        console.log(`📢 [${topic}] ${payload.user}: ${payload.message}`);
      } else {
        console.log(`🔔 [${topic}] ${JSON.stringify(payload)}`);
      }
      process.stdout.write(`[${userName}] > `);
    } catch (_) {}
  }
}

// Função helper para enviar requests via REQ/REP
async function sendRequest(reqSocket, service, data = {}) {
  incClock();
  const request = { service, data: { ...data, lamport_clock: lamportClock } };
  await reqSocket.send(JSON.stringify(request));
  const [replyBuf] = await reqSocket.receive();
  let reply = {};
  try {
    reply = JSON.parse(replyBuf.toString());
  } catch {
    reply = replyBuf.toString();
  }
  updateClock(reply.lamport_clock || 0);
  return reply;
}

// Menu principal (login e registro)
async function main() {
  const reqSocket = new zmq.Request();
  reqSocket.connect(REQ_ADDR);

  while (true) {
    console.log("\n--- CLIENTE INTERATIVO ---");
    console.log("1. Login");
    console.log("2. Cadastrar novo usuário");
    console.log("3. Sair");
    const choice = prompt("Escolha: ");

    incClock();
    let request = { data: { lamport_clock: lamportClock } };

    if (choice === '1' || choice === '2') {
      request.data.user = prompt("Usuário: ");
      request.data.senha = prompt("Senha: ", { echo: '' });
      request.service = choice === '1' ? 'login' : 'addUser';

      try {
        await reqSocket.send(JSON.stringify(request));
        const [replyBuf] = await reqSocket.receive();
        const res = JSON.parse(replyBuf.toString());
        updateClock(res.lamport_clock || 0);

        if (choice === '1') {
          if (res.status === "OK") {
            console.log("✅ Login realizado com sucesso!");
            currentUser = res.user;
            sessionToken = res.token;
            isBot = currentUser.startsWith("bot-");
            messageListener(currentUser, isBot).catch(e => console.error("Listener error:", e));
            await mainApp(currentUser, sessionToken, reqSocket);
            break;
          } else console.log(`❌ Falha: ${res.message}`);
        } else {
          if (res.status === "OK") console.log("✅ Usuário cadastrado! Faça login agora.");
          else console.log(`❌ ${res.message}`);
        }
      } catch (e) {
        console.error("Erro na comunicação com o servidor:", e);
      }

    } else if (choice === '3') {
      console.log("Saindo...");
      process.exit(0);
    } else {
      console.log("Opção inválida.");
    }
  }
}

// Função pós-login
async function mainApp(user_nome, session_token, reqSocket) {
  await setTimeout(200);

  console.log("\n--- Bem-vindo ao sistema ---");
  console.log("Digite 'ajuda' para ver os comandos disponíveis.\n");

  while (true) {
    const opcao = prompt(`[${user_nome}] > `);
    if (!opcao) continue;
    const [command, ...args] = opcao.toLowerCase().split(' ');

    incClock();
    let request = { data: { token: session_token, user: user_nome, lamport_clock: lamportClock } };

    if (command === 'sair') {
      console.log("👋 Encerrando...");
      process.exit(0);

    } else if (command === 'ajuda') {
      console.log("\nComandos disponíveis:");
      console.log("listar_canais            - Lista canais");
      console.log("listar_user              - Lista usuários");
      console.log("add_canal                - Cria novo canal");
      console.log("inscrever <canal>        - Inscreve-se num canal");
      console.log("publicar <canal>         - Publica mensagem num canal inscrito");
      console.log("mensagem <user>          - Envia mensagem direta");
      console.log("sync_time                - Sincroniza relógio");
      console.log("sair                     - Fecha o cliente");
      console.log("ajuda                    - Mostra este menu\n");

    } else if (command === 'listar_canais' || command === 'listar_user' || command === 'sync_time') {
      request.service = command === 'listar_canais' ? 'listChannels' :
                        command === 'listar_user' ? 'listUsers' : 'getTime';

      try {
        const res = await sendRequest(reqSocket, request.service, request.data);
        if (command === 'listar_canais') {
          console.log("\n--- Canais ---");
          Object.values(res).forEach(c => c.titulo && console.log("→", c.titulo));
        } else if (command === 'listar_user') {
          console.log("\n--- Usuários ---");
          Object.values(res).forEach(u => u.user && console.log("→", u.user));
        } else {
          console.log("\n--- Relógio ---");
          console.log(`Servidor: ${res.server_time_utc}`);
          console.log(`Local:    ${new Date().toISOString()}`);
        }
      } catch (e) { console.error("Erro:", e); }

    } else if (command === 'add_canal') {
      request.service = 'addChannel';
      request.data.titulo = prompt("Nome do canal: ").toLowerCase();
      request.data.desc = prompt("Descrição: ");
      const res = await sendRequest(reqSocket, 'addChannel', request.data);
      if (res.status === "OK") {
        console.log("✅ Canal criado!");
        // bots se inscrevem automaticamente
        if (isBot && subSocket) subSocket.subscribe(request.data.titulo);
      } else console.log("❌", res.message);

    } else if (command === 'inscrever') {
      const canal = args[0];
      if (!canal) return console.log("❌ Use: inscrever <canal>");
      const res = await sendRequest(reqSocket, 'subscribe', { ...request.data, channel: canal });
      if (res.status === "OK") {
        console.log(`✅ ${res.message}`);
        if (subSocket) subSocket.subscribe(canal);
      } else console.log(`❌ ${res.message}`);

    } else if (command === 'publicar') {
      const canal = args[0];
      if (!canal) return console.log("❌ Use: publicar <canal>");
      const mensagem = prompt("Mensagem: ");
      const res = await sendRequest(reqSocket, 'publish', {
        ...request.data,
        channel: canal,
        message: mensagem,
        timestamp: new Date().toISOString()
      });
      if (res.status === "OK") console.log("✅ Mensagem publicada!");
      else console.log("❌", res.message);

    } else if (command === 'mensagem') {
      const dst = args[0];
      if (!dst) return console.log("❌ Use: mensagem <user>");
      const mensagem = prompt("Mensagem: ");
      const res = await sendRequest(reqSocket, 'message', {
        ...request.data,
        dst,
        message: mensagem,
        timestamp: new Date().toISOString()
      });
      if (res.status === "OK") console.log("✅ Mensagem enviada!");
      else console.log("❌", res.message);

    } else {
      console.log("Comando inválido. Digite 'ajuda'.");
    }
  }
}

// Início
main().catch(e => console.error("Erro fatal:", e));
